[quant][fx][bug] Fix error in convert step for QAT

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45050 [quant][fx][bug] Fix error in convert step for QAT**

Summary:
Update tests to actually test for QAT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23808022](https://our.internmc.facebook.com/intern/diff/D23808022)

supriyar committed:
[quant][fx][bug] Fix error in convert step for QAT
Summary:
Update tests to actually test for QAT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx][bug] Fix error in convert step for QAT"


Summary:
Update tests to actually test for QAT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23808022](https://our.internmc.facebook.com/intern/diff/D23808022)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -511,7 +511,7 @@ def load_arg(a):
         to_be_removed = []
         for name, module in model.named_modules():
-            if is_activation_post_process(module):
+            if 'activation_post_process_' in name.split('.')[-1]:
```
why is this changed?

supriyar replied:
This is for the observer modules that are an attr of the fakequant module. If we just check the module type then it returns true and then fails when we try to call `delattr` below.

jerryzh168 replied:
I see, maybe make more sense to fix this by avoiding checking the submodule of fakequantize module?

supriyar replied:
The submodule (i.e observer) shows up in `model.named_modules()` as well. So we need some additional check to not add that in the removed list.

jerryzh168 replied:
yes, I think so, feel free to move this to a function

supriyar committed:
Update on "[quant][fx][bug] Fix error in convert step for QAT"


Summary:
Update tests to actually test for QAT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23808022](https://our.internmc.facebook.com/intern/diff/D23808022)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -126,6 +126,12 @@ def is_activation_post_process(module):
             isinstance(module, torch.quantization.FakeQuantize))
 
+def is_submodule_of_fake_quant(name, module, named_modules):
+    parent_name = name.rsplit('.', 1)
+    if len(parent_name) > 1:
+        return is_activation_post_process(named_modules[parent_name[0]])
```
nit: we have a _parent_name helper function I think, you can use that

jerryzh168 reviewed (approved):
Looks good, please use _parent_name

supriyar committed:
Update on "[quant][fx][bug] Fix error in convert step for QAT"


Summary:
Update tests to actually test for QAT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23808022](https://our.internmc.facebook.com/intern/diff/D23808022)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx][bug] Fix error in convert step for QAT"


Summary:
Update tests to actually test for QAT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_linear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23808022](https://our.internmc.facebook.com/intern/diff/D23808022)

[ghstack-poisoned]
